### PR TITLE
PAYARA-4138 Hot deploy the Engine of application

### PR DESCRIPTION
**Pull request by jGauravGupta**
# Description
This is a feature that improves deployment performance in `--hotdeploy` mode with the following operations:
.
- If `deploy` command invoked with `--hotdeploy=true` flag and application not already deployed then Application info (`EngineInfo`, `ModuleInfo`, Dol Processor result and `DeploymentContext` properties, etc) stored in `ApplicationState`.
- If `redeploy` command invoked with `--hotdeploy=true` and `--metadataChanged=false` then all previously-stored objects are reused, classloader replaced (although classes binary content copied from the previous classloader) and `DolProcessor`, `WebBundleDescriptor` processing, `undeploy` operation, write operation in domain.xml and many other operations are skipped that improves redeployment performance up to 50% compared to normal redeployment.
- If `redeploy` command invoked with `--hotdeploy=true` and `--metadataChanged=true` then all previously-stored objects are removed (staled), usual `undeploy` operation performed and a new application state is cached.
- `--metadataChanged` flag is set to true whenever application descriptor, application resource, meta-inf file modified or/and classes file removed from IDE.

## Features
- ApplicationState (`ModuleInfo`, `EngineInfo`, `DeploymentContext`, etc) cached https://github.com/payara/Payara/pull/4363/commits/49544478822d1bc15130211b70cda743f3007ee1
- Dynamic DOL Processor https://github.com/payara/Payara/pull/4363/commits/2f81c775bc769a605919e631c982482d98900361 https://github.com/payara/Payara/pull/4363/commits/6bef052e0b2ba523d04b027ff578f950841a821d https://github.com/payara/Payara/pull/4363/commits/952cfb55c678557d664117b4549a74c7530840ff
- Servlet Annotation Handler optimization https://github.com/payara/Payara/pull/4363/commits/8b7ec67cbad2fd0c302edd3fca7e224fd6b3b13e
- EJB Annotation Handler optimization https://github.com/payara/Payara/pull/4363/commits/166fd8c960cdbbdbcf502667f8c6c25cf4643e3c
- Reuse `binaryContent` of previous classloader's classes https://github.com/payara/Payara/pull/4363/commits/ea5904cdd6dcac2a58e0b334b32ef0ebc2afb8b1 https://github.com/payara/Payara/pull/4363/commits/8f9163e2001fb74b983bc11ec66d877cd37cb2d1 https://github.com/payara/Payara/pull/4363/commits/810cd3e26b0c150114b93947744106a632cf018d

# Important Info

### Dependant PRs 
https://github.com/apache/netbeans/pull/1562

# Testing

### Testing Performed
Manual testing (specific Arquillian extension needed to be designed for the automated tests)

### Testing steps
- Download the NetBeans IDE build from [here](https://drive.google.com/open?id=1Mn7WByydrgac0e62xDKvgM9yi2dYZARw)
- Open the IDE and add the Payara Server.
- Open the properties panel of the Payara Server and enable the `Hot Deploy` checkbox.
- Create Maven Web Application and make sure `compile on save` and `deploy on save` options are enabled in application properties.
- Now deploy the maven web application to Payara Server.
- Open any java class from the project to update it and save.
- For real-life applications with dozens of servlet/rest & other java classes, by enabling the `Hot Deploy` mode, you may observe 45-50% performance improvement in redeploy operation compare to `vanilla` redeployment.

## Notes for Reviewers
- Currently, EJB timer service throws an error during hot deployment operation which will be fixed with partial start/stop operation of container: https://payara.atlassian.net/browse/ECOSYS-90.

### CODE REVIEW

File: nucleus/deployment/admin/src/main/java/org/glassfish/deployment/admin/DeployCommand.java [hunk 1]
```diff
@@ -492,8 +492,6 @@ public void execute(AdminCommandContext context) {
             deploymentContext
                     = deployment.getBuilder(logger, this, report).
                             source(initialContext.getSource()).archiveHandler(archiveHandler).build(initialContext);
-            structuredTracing.register(deploymentContext);
```
Comment (jGauravGupta) [inline nucleus/deployment/admin/src/main/java/org/glassfish/deployment/admin/DeployCommand.java]: `structuredTracing` already registered as `initialContext` and `deploymentContext` are same instance.

**Comment by jGauravGupta** [comment]
jenkins test please

**Comment by jGauravGupta** [comment]
jenkins test please

**Comment by jGauravGupta** [comment]
jenkins test please

File: nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationState.java [hunk 2]
```diff
@@ -132,27 +136,80 @@ public void setModuleInfo(ModuleInfo moduleInfo) {
...
+                .forEach(e -> newContext.addTransientAppMetaData(e.getKey(), e.getValue()));
+        this.deploymentContext = newContext;
+        this.previousClassLoaders = getClassLoaders(this.applicationInfo);
+        events.send(
+                new EventListener.Event<ApplicationInfo>(
+                        Deployment.APPLICATION_UNLOADED,
+                        this.applicationInfo
+                ),
+                false
+        );
```
Comment (jGauravGupta) [inline nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/ApplicationState.java]: `APPLICATION_UNLOADED` event snippets will be removed under https://payara.atlassian.net/browse/ECOSYS-90 partial start/stop of the container.

**Comment by Pandrex247** [comment]
There is a conflict that needs resolving though ;)
It's just git being git because someone else touched the file

**Comment by jGauravGupta** [comment]
jenkins test please

**Comment by MarkWareham** [comment]
Poke to merge? @jGauravGupta

**Pull request merged**